parlai/tasks/reasoning/reason_types/step_by_step.py

```diff
@@ -0,0 +1,955 @@
+            swapped_word_step = self.swap_words(step)
+            if swapped_word_step != "" and str(step) != swapped_word_step:
+                grammatical_error_steps.append((i, swapped_word_step))
+
+        if not grammatical_error_steps:
+            raise RuntimeError(
+                "GrammaticalErrorStep failed for the reasoning chain "
+                + " ".join(str(step) for step in steps)
+                + ". Make sure get_data_for_fold()` appropriately filters for these turns when used with this perturbation."
+            )
+
+        to_dupe = self.random.choice(range(len(grammatical_error_steps)))
+        result[grammatical_error_steps[to_dupe][0]] = grammatical_error_steps[to_dupe][
+            1
+        ]
+        example_dict['steps'] = result
+        return example_dict
+
+
+class NegateStep(StepPerturbation):
+    """
+    Negates one step in example.
+
+    Dependencies:
+    pip3 install checklist
+    python -m spacy download en_core_web_sm
+    """
+
+    def __init__(self, opt: Opt, cache: Optional[Dict[str, List[List[str]]]] = None):
+        super().__init__(opt, cache)
+        self.nlp = spacy.load('en_core_web_sm')
+
+    def perturb(self, example_dict: Dict) -> Dict:
+        example_dict = copy.deepcopy(example_dict)
+        steps = example_dict['steps']
+        if len(steps) == 0:
+            raise RuntimeError(
+                "NegateSteps requires at least 1 step. Make sure `get_data_for_fold`()`"
+                + "appropriately filters for these turns when used with this perturbation."
+            )
+        result = copy.deepcopy(steps)
+        negated_steps = []
+        for i, step_to_dupe in enumerate(result):
+            pdata = list(self.nlp.pipe([str(step_to_dupe)]))
+            try:
+                ret = Perturb.perturb(pdata, Perturb.add_negation)
+                if ret.data:
+                    negated_steps.append((i, ret.data[0][1]))
+            # Perturb.add_negation is noted as still experimental in documentation - occasionally
+            # fails on certain cases with a Runtime or Index Error
+            except RuntimeError:
+                pass
+            except IndexError:
+                pass
+        if not negated_steps:
+            raise RuntimeError(
+                "NegateStep failed for the reasoning chain "
+                + " ".join(str(step) for step in steps)
+                + ". Make sure get_data_for_fold()` appropriately filters for these turns"
+                + "when used with this perturbation."
+            )
+        to_dupe = self.random.choice(range(len(negated_steps)))
+        result[negated_steps[to_dupe][0]] = negated_steps[to_dupe][1]
+        example_dict['steps'] = result
+        return example_dict
+
+
+class SemanticChangeStep(StepPerturbation):
+    """
+    Change the semantics of one step in example by replacing entities.
+
+    Dependencies:
+    pip3 install checklist
```

Comment:
It doesn't look like `SemanticChangeStep` uses checklist? Perhaps this got copied over with the "spacy" comment, since it _does_ depend on spacy. (or maybe I'm missing something)

Reply:
Yep, don't see it either